libraries/chain/block_log.cpp

```diff
@@ -1475,4 +1508,238 @@ namespace eosio { namespace chain {
+      fseek(ind_in, -sizeof(uint64_t), SEEK_END);
+      fread((void*)&index_pos, sizeof(index_pos), 1, ind_in);
+
+      // Find number of indexes in index file. To be used in
+      // later functions.
+      fseek(ind_in, 0, SEEK_END);
+      uint64_t num_indexes_existing = ftell(ind_in) / sizeof(uint64_t);
+
+      // block_pos should be the same as index_pos
+      if (index_pos > block_pos) {
+         ilog("Index file has extra entries, truncate it");
+         trim_index(block_pos, num_indexes_existing);
+      } else if (index_pos < block_pos) {
+         ilog("Index file is incomplete, append missing indexes to it");
+         append_index(num_indexes_existing);
+      }
+   }
+
+   // Trim index file until its last index matches head block's
+   // position in block file
+   void trim_data::trim_index(uint64_t head_block_pos, uint64_t num_indexes_existing) {
+      // It is much faster to traverse index file than traverse
+      // blocks in block file
+
+      // Starting from last index in index file, compare it with
+      // head_block_pos. If matches, truncate index file from there.
+      for (auto  i = 1; i <= num_indexes_existing; ++i) {
+         // get the location storing the last block's index
+         // please note the negative sign and multiplication by i
+         auto status = fseek(ind_in, -sizeof(uint64_t) * i, SEEK_END);
+
+         // read last block's position (index)
+         uint64_t block_pos_by_index = 0;
+         fread((void*)&block_pos_by_index, sizeof(uint64_t), 1, ind_in);
+
+         if (block_pos_by_index == head_block_pos) {
+            // Found the index matching the last good block.
+            // Truncate all indexes after the matching index.
+            boost::filesystem::resize_file(index_file_name, sizeof(uint64_t) * (num_indexes_existing - i + 1));
+            return;
+         }
+      }
+
+      // Either block file or index file is corrupt beyond fixable
+      // Assert it.
+      EOS_ASSERT(false, block_log_exception, "Index file has extra index, but cannot be trimmed due to no matching index of head block found in index file");
+   }
+
+   // When we already have an index file, it will be much faster to
+   // append missing indexes than build from the scratch
+   void trim_data::append_index(uint64_t num_indexes_existing) {
+      detail::reverse_iterator block_log_iter;
+
+      const uint32_t num_blocks = block_log_iter.open(block_file_name);
+      EOS_ASSERT(num_blocks > num_indexes_existing, block_log_exception, "Number of blocks in block file ${num_blocks} must be greater than blocks in index file ${num_indexes_existing}", ("num_blocks",num_blocks)("num_indexes_existing", num_indexes_existing));
+
+      uint32_t num_indexes_to_append = num_blocks - num_indexes_existing;
+
+      // Create a temporary file to store missing indexes
+      auto now = fc::time_point::now();
+      fc::path temp_file_name = index_file_name.generic_string().append("-").append( now );
```

Comment:
You can use ```fc::temp_file temp_file;```.  This will automatically release the file after it goes out of scope and creates it in the system temporary directory.

Reply:
Agreed. Thanks for pointing it out. It simplifies the code.